legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabUserContext.java

```diff
@@ -65,46 +65,83 @@ public GitLabApi getGitLabAPI(boolean redirectAllowed)
         {
             GitLabSession gitLabSession = getGitLabSession();
             GitLabToken token = gitLabSession.getGitLabToken();
+            GitLabTokenResponse tokenResponse;
             if (token == null)
             {
-                try
-                {
-                    token = authorizerManager.authorize(session, this.appInfo);
-                }
-                catch (GitLabOAuthAuthenticator.UserInputRequiredException e)
-                {
-                    URI redirectURI = GitLabOAuthAuthenticator.buildAppAuthorizationURI(e.getAppInfo(), this.httpRequest);
-                    throw new LegendSDLCServerException(redirectURI.toString(), Status.FOUND);
-                }
-                catch (GitLabAuthFailureException e)
-                {
-                    throw new LegendSDLCServerException(e.getMessage(), Status.FORBIDDEN, e);
-                }
-                catch (GitLabAuthException e)
-                {
-                    throw new LegendSDLCServerException(e.getMessage(), Status.INTERNAL_SERVER_ERROR, e);
-                }
-                if (token != null)
-                {
-                    gitLabSession.setGitLabToken(token);
-                    LegendSDLCWebFilter.setSessionCookie(this.httpResponse, gitLabSession);
-                }
-                else if (redirectAllowed)
+                token = setGitlabTokenForSession(redirectAllowed, gitLabSession);
+            }
+            else if (gitLabSession.shouldRefreshToken())
+            {
+                if (gitLabSession.getRefreshToken() != null)
                 {
-                    URI redirectURI = GitLabOAuthAuthenticator.buildAppAuthorizationURI(this.appInfo, this.httpRequest);
-                    throw new LegendSDLCServerException(redirectURI.toString(), Status.FOUND);
+                    try
+                    {
+                        tokenResponse = GitLabOAuthAuthenticator.getOAuthTokenFromRefreshToken(gitLabSession.getRefreshToken(), appInfo);
+                        if (tokenResponse != null)
+                        {
+                            gitLabSession.setGitLabToken(tokenResponse.getAccessToken());
+                            gitLabSession.setRefreshToken(tokenResponse.getRefreshToken());
+                            gitLabSession.setTokenExpiry(tokenResponse.getExpiresInSecs());
+                            token = gitLabSession.getGitLabToken();
+                            LegendSDLCWebFilter.setSessionCookie(this.httpResponse, gitLabSession);
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        token = setGitlabTokenForSession(redirectAllowed, gitLabSession);
+                    }
```

Comment:
We might want to log the exception here, at least in some cases. We definitely should log it if it's not the kind of exception we'd expect from refreshing the token. I don't know whether it's worth logging in other cases.

Reply:
I agree maybe we can log gitlabAuthExceptions (things coming from gitlab) and just let any other exception not be logged

Reply:
Let's log all exceptions with level warn: LOGGER.warn("Error refreshing token", e).